### PR TITLE
Enhancement: adds support for conformant, visually-hidden table headers

### PR DESCRIPTION
**Pull request by MelSumner**
This PR adds the functionality for an accessible but visually hidden column name.

### CODE REVIEW

File: packages/components/addon/components/hds/table/index.hbs [hunk 1]
```diff
@@ -22,16 +22,22 @@
               {{column.label}}
             </Hds::Table::ThSort>
           {{else}}
-            <Hds::Table::Th @align={{column.align}} @width={{column.width}}>{{column.label}}</Hds::Table::Th>
+            <Hds::Table::Th @align={{column.align}} @width={{column.width}}>
+              {{#if this.isHidden}}
```
Comment (alex-ju) [inline packages/components/addon/components/hds/table/index.hbs:26]: if the API is going to be

```
@columns={{array
      (hash key="artist" label="Artist" isSortable="true")
      (hash key="album" label="Album" isSortable="true")
      (hash key="year" label="Release Year" isHidden="true")
```
then the only required change would the following line
```suggestion
              {{#if column.isHidden}}
```

Comment (MelSumner) [reply packages/components/addon/components/hds/table/index.hbs]: Ahh yes that does it! 
CleanShot 2023-07-31 at 10 26 54@2x

**Comment by MelSumner** [comment]
@hashicorp/hds-design @hashicorp/hds-engineering do we want to ship this consumer-requested feature or put it on hold?

**Comment by didoo** [comment]
> @hashicorp/hds-design @hashicorp/hds-engineering do we want to ship this consumer-requested feature or put it on hold?

Since it was in draft mode, I didn't look into it. I'll have a look tomorrow.

**Comment by MelSumner** [comment]
@didoo yeah I figured it was a way for us to make a decision one way or another. Either is fine but we should decide.

**Review by didoo — COMMENTED** [review]
I have added a bunch of comments (technically blockers, but since it's in the PR is in early stage I imagine the code was not complete/fully reviewed).

More in general, I am not convinced by the APIs for this feature.
We're exposing the `isHidden` at both the `@columns` level and the `Th/ThSort` level, but I don't think we need to have it for the `ThSort` (no real use case) and this simplifies the logic a bit; with this simplification in mind, we can think of a `Th` with `@isHidden` + `isHidden` key for the `@columns`.
The problem I see is that the `Th` element is used not only for column headers, but also for row headers and it can contain structured content (see for example here: https://github.com/hashicorp/cloud-ui/blob/main/engines/waypoint/addon/components/resources-table.hbs#L28-L39)
which would lead to an API that would allow something like this (even it it doesn't make sense):
```
<B.Th @isHidden={{true}} data-test-id-column class="waypoint-table__id-column">
  <Hds::Link::Inline
    title={{bundle.resource.name}}
    @route="projects.project.apps.app.resources.resource"
    @model={{bundle.resource.id}}
    @color="secondary"
  >
    {{bundle.resource.name}}
  </Hds::Link::Inline>
</B.Th>
```
and more in general we could have a `<span>` wrapping unknown HTML elements.

So I am proposing an alternative API:
- we introduce a `ThHidden` subcomponent which is very simple (essentially the `Th` subcomponent with a subset of arguments)
- in the `columns` loop ([here](https://github.com/hashicorp/design-system/blob/main/packages/components/addon/components/hds/table/index.hbs#L14)) we introduce another condition `{{#if column.isHidden}}` (argument name TBD, `isHidden` is too generic) that if satisfied renders the `ThHidden` subcomponent with the `{{column.label}}` value as text)
- consumers can use the `ThHidden` whenever they need to build custom tables not relying on the `@columns/@model` looping)
- we add the documentation for this new sub-component on the website, without impacting the existing APIs documentation (otherwise we have to explain what to do and what not to do, with the current proposal)

**Comment by MelSumner** [comment]
> I don't think we need to have it for the `ThSort` (no real use case)

So using it in conjunction with `ThSort` was exactly the request from the user; see our internal slack chat/channel for context.

File: packages/components/addon/components/hds/table/th-sort.hbs [hunk 2]
```diff
@@ -9,10 +9,14 @@
   ...attributes
   scope="col"
 >
-  <button type="button" {{on "click" this.onClick}}>
-    <div class="hds-table__th-sort--button-content">
-      <span class="hds-typography-body-200 hds-font-weight-semibold">{{yield}}</span>
-      <FlightIcon @name={{this.icon}} />
-    </div>
-  </button>
+  {{#if this.isHidden}}
```
[the remark this answers is not in the document]
Comment (MelSumner) [reply packages/components/addon/components/hds/table/th-sort.hbs:12]: It's a use case where some columns are sortable and one column header is hidden.

**Comment by didoo** [comment]
> > I don't think we need to have it for the `ThSort` (no real use case)
> 
> So using it in conjunction with `ThSort` was exactly the request from the user; see our internal slack chat/channel for context.

In conjunction **with** yes, of course. But not that the table column header is **both sortable and hidden**.

File: packages/components/addon/components/hds/table/th-sort.hbs (hunk 2 above)
[the remark this answers is not in the document]
Comment (didoo) [reply packages/components/addon/components/hds/table/th-sort.hbs:12]: See comment above. The point is that the column header which is visually hidden is **not** also sortable.

**Comment by MelSumner** [comment]
> In conjunction with yes, of course. But not that the table column header is both sortable and hidden.

Yes, correct. 

I'm going to go ahead and close this PR and open a JIRA ticket. I thought maybe we could quickly ship a useful feature for a consumer since we have some precedence for this, but it's clear we need more record of context and prioritization. 

Thanks for taking a look at this PR and providing feedback.

**Pull request closed without merging**